### PR TITLE
Test resourcetrail

### PR DESCRIPTION
**Pull request by mastrHyperion98**
Okay so this is a big one, essentially we are charging algorithm after discovering a couple erroneous edge cases. Feel free to test a variety of scenarios to make sure the collected information is accurate. Significant changes have been made in many various areas. 

<h1>Changes</h1>

- Switch algorithm to ResourceTracker. Functions names have been renamed and data structure modified such that it remained the same as ResourceScore. It shouldnt affect the current code base. If anything we will need to rename a thing or two. 

- ResourceTrails is no longer directional, it is now birdirectional.

- Except for the root every node in the Resource Trails may have both incoming and outgoing edges. and be connected with ALL adjacent tiles other than the root of the trail. 

- ScoreDriver now has three test. The one made by Mahdi that originally both our algorithms failed. The loop and non loop test I originally designed

- Fixes the indices inversion for the ResourceTracker

- Fixes all memory allocation/deallocation and copying issues

- Refactor to optimize and shrink the code as much as possible W/O altering its current logic.

- Increase MAX_VISIT for the ResourceTracker Quads to 100. In worse case the efficiency of the algorithm is in theory 100*N. However, this is just by safety net to prevent any possible lock up or infinite loop which "could" in theory occur otherwise. This only restricts the number of times a node can be added to the queue. Thus limiting the number of iterations for the algorithm. The number a times an actual tile can be visited is limitless. Analytically i dont see any evidence to show a direct relationship between the number of nodes and the amount of iterations needed. For now it works. I will continue to look into optimizing the code to bring it down closer to O(N) as right now it stands at O(N^2). For now it is functional and fast enough. I have a couple idea in mind to optimize and speed up the algorithm. However, I will approach these in a separate branch. 

If you see some optimization idea of your own while reviewing the code, please let me know and I will address them in the aforementioned branch. 
 
I will only merge once I have at least two approvals!

### CODE REVIEW

File: NewHaven/src/GBMap.h [hunk 1]
```diff
@@ -33,8 +33,9 @@ class GBMap{
     void printIndexConfiguration();
     bool addBuildingToBoard(Building &building);
     Building* drawBuildingFromBoard(int position);
-private:
     GameBoard* board;
+private:
+
```
Comment (MahChaa) [inline NewHaven/src/GBMap.h:38]: I had made the board public for testing purposes, does need to stay that way?

File: NewHaven/src/Player.cpp [hunk 2]
```diff
@@ -2,14 +2,14 @@
...
     vb_score = new ScoreCalculator();
     my_hand = new Hand();
 }
 
 Player::Player(const Player &player) {
     village = new VGMap(*player.village);
-    resource_score = new ResourceScore(*player.resource_score);
+    //resource_score = new ResourceScore(*player.resource_score);
     vb_score = new ScoreCalculator(*player.vb_score);
     my_hand = new Hand(*player.my_hand);
```
Comment (MahChaa) [inline NewHaven/src/Player.cpp]: You don't copy the resourceTracker and you also don't initialize it. You need to pick one or the other.
Also there's unnecessary copying. Using the copy constructor and the copy assignment at the same time isn't efficient.

File: NewHaven/src/Player.h [hunk 3]
```diff
@@ -17,13 +17,13 @@ struct Player
     // cannot implement yet. Requires the game_controller / main-loop / logic controller to be a singleton with references to the harvest tile
     void drawHarvestTile(HarvestTile& tile);
     // im assuming this prints out the available resources;
-    ResourceScore resourceTracker();
+    ResourceTracker resourceTracker();
     bool buildVillage();
```
Comment (MahChaa) [inline NewHaven/src/Player.h]: `ResourceTracker getResourceTracker() const;`
Is more readable, I had to look at the implementation to understand what the function did, while you could just make it clear in the definition.

File: NewHaven/src/ResourceTracker.cpp [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+               ResourceTypes resource = r_resources[i];
+               addResources(resource);
+                quad->isMatching[i] = true;
+            }
+            (*quad->current_visit_count)++;
+            map.insert(Pair(root, quad));
+    // delete r_resources
+    delete r_resources;
+    while(!queue.empty()){
+        root = queue.front();
```
Comment (MahChaa) [inline NewHaven/src/ResourceTracker.cpp]: Forgotten comment.

File: NewHaven/src/ResourceTracker.cpp [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+        Quad *root_quad = map[root];
+        // now we build the queues
+        ResourceTrails::adjacency_iterator neighbourIt, neighbourEnd;
+        tie(neighbourIt, neighbourEnd) = adjacent_vertices(root, trail);
+        ResourceTypes *root_resources = (*trail[root].tile).getTileContent();
+        for (; neighbourIt != neighbourEnd; ++neighbourIt) {
+            // get ourselves our vertex
```
Comment (MahChaa) [inline NewHaven/src/ResourceTracker.cpp]: You can put the tie statement in the loop to make it more understandable and local to the loop.

File: NewHaven/src/Resources.h [hunk 6]
```diff
@@ -10,10 +10,10 @@ std::ostream& operator<<(std::ostream& output, const ResourceTypes& resource); /
 struct HarvestTile
 {
 private:
-	ResourceTypes* tileContent;	//Array of resource types
 
-public:
 
+public:
+    ResourceTypes* tileContent;	//Array of resource types
```
Comment (MahChaa) [inline NewHaven/src/Resources.h]: Tile content was made public for testing purposes, make it private again when the testing is done.

**Review by MahChaa — CHANGES_REQUESTED** [review]
Remove the unnecessary files like the ResourceScore and edit the CMakeLists.txt accordingly. Your algorithm was correct in the testing (4 different resourceTrails). No need to try to improve efficiency as long as it works for now.

File: NewHaven/src/GBMap.h (hunk 1 above)
Comment (mastrHyperion98) [reply NewHaven/src/GBMap.h:38]: shit forgot about those, no indeed there is no reason to keep them public.

File: NewHaven/src/Player.h (hunk 3 above)
Comment (mastrHyperion98) [reply NewHaven/src/Player.h]: resourceTracker is the name the teacher wants in the assignment unfortunately. I share your opinion on its readability.

File: NewHaven/src/Player.cpp (hunk 2 above)
Comment (mastrHyperion98) [reply NewHaven/src/Player.cpp]: that is my bad i had commented it out for testing and completely forgot to add one for ResourceTracker ooops lol

File: NewHaven/src/Player.cpp [hunk 7]
```diff
@@ -2,14 +2,14 @@
...
     my_hand = new Hand();
 }
 
 Player::Player(const Player &player) {
     village = new VGMap(*player.village);
-    resource_score = new ResourceScore(*player.resource_score);
+    resource_score = new ResourceTracker(*player.resource_score);
     vb_score = new ScoreCalculator(*player.vb_score);
     my_hand = new Hand(*player.my_hand);
 }
```
Comment (MahChaa) [inline NewHaven/src/Player.cpp]: The resource tracker is initialized, but there's still unnecessary copying, use list initialization.
ex: `resource_score{new ResourceTracker(*player.resource_score),`
Also why did we switch to snake case variable naming?

**Review by MahChaa — CHANGES_REQUESTED** [review]

File: NewHaven/src/Player.cpp (hunk 7 above)
Comment (mastrHyperion98) [reply NewHaven/src/Player.cpp]: I've always use snake case for variable names. is that an issue?

**Review by MahChaa — APPROVED** [review]
Perfect, all changes were made successfully. Everything seems to work fine.
